src/parse.rs

```diff
@@ -0,0 +1,2602 @@
+                        SrcSpan { start, end },
+                    )
+                } else {
+                    parse_error(ParseErrorType::ExpectedName, SrcSpan { start, end })
+                }
+            }
+            None => parse_error(ParseErrorType::UnexpectedEOF, SrcSpan { start: 0, end: 0 }),
+        }
+    }
+
+    // Expect an UpName else a token dependent helpful error
+    fn expect_upname(&mut self) -> Result<(usize, String, usize), ParseError> {
+        let t = self.next_tok();
+        match t {
+            Some((start, tok, end)) => {
+                if let Tok::Name { .. } = tok {
+                    parse_error(ParseErrorType::IncorrectUpName, SrcSpan { start, end })
+                } else if let Tok::UpName { name } = tok {
+                    Ok((start, name, end))
+                } else if let Tok::DiscardName { .. } = tok {
+                    parse_error(ParseErrorType::IncorrectUpName, SrcSpan { start, end })
+                } else {
+                    parse_error(ParseErrorType::ExpectedUpName, SrcSpan { start, end })
+                }
+            }
+            None => parse_error(ParseErrorType::UnexpectedEOF, SrcSpan { start: 0, end: 0 }),
+        }
+    }
+
+    // Expect a String else error
+    fn expect_string(&mut self) -> Result<(usize, String, usize), ParseError> {
+        match self.next_tok() {
+            Some((start, Tok::String { value }, end)) => Ok((start, value, end)),
+            _ => self.next_tok_unexpected(vec!["a string".to_string()]),
+        }
+    }
+
+    // If the next token matches the requested, consume it and return (start, end)
+    fn maybe_one(&mut self, tok: &Tok) -> Option<(usize, usize)> {
+        match self.tok0.take() {
+            Some((s, t, e)) if t == *tok => {
+                self.next_tok();
+                Some((s, e))
+            }
+
+            t0 => {
+                self.tok0 = t0;
+                None
+            }
+        }
+    }
+
+    // Parse a series by repeating a parser, and possibly a separator
+    fn series_of<A>(
+        &mut self,
+        parser: &dyn Fn(&mut Self) -> Result<Option<A>, ParseError>,
+        sep: Option<&Tok>,
+    ) -> Result<Vec<A>, ParseError> {
+        let mut results = vec![];
+        while let Some(result) = parser(self)? {
+            results.push(result);
+            if let Some(sep) = sep {
+                if self.maybe_one(sep).is_none() {
+                    break;
+                }
+                // Helpful error if extra separator
+                if let Some((start, end)) = self.maybe_one(sep) {
+                    return parse_error(ParseErrorType::ExtraSeparator, SrcSpan { start, end });
+                }
+            }
+        }
+
+        Ok(results)
+    }
+
+    // If next token is a Name, consume it and return relevant info, otherwise, return none
+    // don't error in here. Result type is for combinator ease
```

Comment:
" Result type is for combinator ease"

When is this useful? From what I can see this is often called with `?` which always unwraps the `Ok`. What am I missing?

Reply:
the `maybe_` functions that get passed to `series_of` return a `Result<Option<>>` rather than just an `Option` as `series_of` requires it. `maybe_binding_start` and `maybe_one` return simply an `Option` as they never get passed to `series_of`. Happy to re-work this. What do you think?

Reply:
Oh yes. I think I would slightly prefer to have them wrapped at the call site.